### PR TITLE
Merging javadoc feature branch changes to main

### LINKED ISSUE #449: Fix precommit check failures with missingJavadoc task
**Describe the bug**
The `missingJavadoc` task in [javadoc branch](https://github.com/opensearch-project/OpenSearch/tree/javadoc) currently works well when run on its own (i.e. `./gradlew missingJavadoc`) but fails the precommit checks due to some wrangled gradle build logic. This can be reproduced by running `./gradlew precommit` on `javadoc` branch. The `validateNebulaPom` task, which is part of precommit checks, fails because it looks for pom file in the wrong location (`build/publications/nebula/pom-default.xml`) when the actual pom file is generated in the path configured [here](https://github.com/opensearch-project/OpenSearch/blob/javadoc/buildSrc/src/main/java/org/opensearch/gradle/PublishPlugin.java#L76) (e.g. for benchmarks module`build/distributions/opensearch-benchmarks-7.10.3-SNAPSHOT.pom`). On removing this [line](https://github.com/opensearch-project/OpenSearch/blob/javadoc/build.gradle#L51), the precommit logic seems to work fine. Somehow, applying missingJavadoc script plugin here is causing the PomValidation task to read pom from default location instead of overridden location.

**To Reproduce**
1. Sync remote `javadoc` branch
2. Run `./gradlew precommit`. It should fail in PomValidation
3. Comment this line.
4. Run step 2 again. It should pass now.

**Expected behavior**
`precommit` task should work fine with `missingJavadoc` gradle task.

**Host/Environment (please complete the following information):**
 - OS: iOS
 - Version: 10.14.6

**Logs**
```
➜  search git:(javadoc) ./gradlew precommit

> Configure project :
========================= WARNING =========================
         Backwards compatibility tests are disabled!
See https://github.com/opensearch-project/OpenSearch/issues/105
===========================================================

> Configure project :qa:os
Cannot add task 'destructiveDistroTest.docker' as a task with that name already exists.
=======================================
OpenSearch Build Hamster says Hello!
  Gradle Version        : 6.6.1
  OS Info               : Mac OS X 10.14.6 (x86_64)
  JDK Version           : 14 (OpenJDK)
  JAVA_HOME             : /Library/Java/JavaVirtualMachines/jdk-14.jdk/Contents/Home
  Random Testing Seed   : C2BC3255C3EA0731
  In FIPS 140 mode      : false
=======================================

> Task :benchmarks:validateNebulaPom FAILED

FAILURE: Build failed with an exception.

* What went wrong:
A problem was found with the configuration of task ':benchmarks:validateNebulaPom' (type 'PomValidationTask').
> File '/Users/setiah/projects/odfe/search/benchmarks/build/publications/nebula/pom-default.xml' specified for property 'pomFile' does not exist.
```

### PR DESCRIPTION
**Pull request by setiah**
### Description
This PR merges changes in javadoc feature branch to main branch. It includes -
1. A new gradle plugin to validate missing javadocs 
2. Fix for precommit task failure due to addition of new plugin. (PR #707)
3. Fix Javadoc errors in module `:client:rest` (PR #685)
4. Updates CONTRIBUTING.md
 
### Issues Resolved
#449 , #221 
 
### Check List
- [ ] New functionality includes testing.
  - [ ] All tests pass
- [ ] New functionality has been documented.
  - [ ] New functionality has javadoc added
- [ ] Commits are signed per the DCO using --signoff 

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

Signed-off-by: Himanshu Setia <setiah@amazon.com>

### CODE REVIEW

File: gradle/missing-javadoc.gradle [hunk 1]
```diff
@@ -0,0 +1,255 @@
+/*
```
Comment (setiah) [inline gradle/missing-javadoc.gradle:1]: self review - change license header

File: buildSrc/reaper/src/main/java/org/elasticsearch/gradle/reaper/package-info.java [hunk 2]
```diff
@@ -0,0 +1,26 @@
+/*
```
Comment (setiah) [inline buildSrc/reaper/src/main/java/org/elasticsearch/gradle/reaper/package-info.java:1]: self review - change license header

File: test/fixtures/hdfs-fixture/build.gradle [hunk 3]
```diff
@@ -30,6 +30,8 @@
 
 apply plugin: 'opensearch.java'
 
+group = 'hdfs'
```
Comment (tlfeng) [inline test/fixtures/hdfs-fixture/build.gradle:33]: Is the name of the group fine? The above 2 names are `fixture.xxx`, but these 2 names without the `fixture` prefix.

**Review by tlfeng — APPROVED** [review]

**Review by gzurowski — APPROVED** [review]
LGTM 👍

**Comment by setiah** [comment]
start gradle check

File: test/fixtures/hdfs-fixture/build.gradle (hunk 3 above)
Comment (setiah) [reply test/fixtures/hdfs-fixture/build.gradle:33]: Yeah, that discrepancy exists from the legacy codebase. The imports inside hdfs module use package name as `hdfs` unlike other fixture modules. Probably we can fix the hdfs module in a separate PR.

**Pull request merged**